tuning.py

```diff
@@ -0,0 +1,308 @@
+    doubles_df['residual'] = residual
+    return doubles_df
+
+
+def get_epistasis_stats(landscape: potts_model.PottsModel,
+                        threshold: float = 0, adaptive: bool = True) -> Tuple[float, float]:
+    """Returns statistics about epistasis for combinations of singles.
+
+    Mean epistasis effect size is defined as the average effect of epistasis when two single mutants
+    from the selected set are combined. The rate of reciprocal sign epistasis is defined as the fraction
+    of epistatic interactions that have effects opposing the effects of the constituent singles. (i.e. the rate
+    of negative epistasis for adaptive singles in combination).
+
+    Args:
+      landscape: The landscape.
+      threshold: The threshold fitness for singles.
+      adaptive: If True, thresholded singles will have fitness >= `threshold`. If False,
+        selected singles will have fitness < `threshold`.
+
+    Returns:
+      A Tuple of length 2 where element [0] is the mean epistasis effect size
+      and element [1] is the rate of reciprocal sign epistasis.
+    """
+    doubles_df = get_doubles_df(landscape, threshold, adaptive)
+    residual = doubles_df.residual
+
+    mean_epistasis_effect = np.mean(residual)
+    if adaptive:
+        rate_reciprocal_epistasis = (residual < 0).sum() / residual.shape[0]
+    else:
+        rate_reciprocal_epistasis = (residual > 0).sum() / residual.shape[0]
+    return mean_epistasis_effect, rate_reciprocal_epistasis
+
+
+def get_mean_single_effect(landscape: potts_model.PottsModel, threshold: float = 0, adaptive: bool = True) -> float:
+    """Returns average effect size of singles for a given threshold.
+
+    Args:
+      landscape: The landscape.
+      threshold: The threshold fitness for singles.
+      adaptive: If True, thresholded singles will have fitness >= `threshold`. If False,
+        selected singles will have fitness < `threshold`.
+
+    Returns:
+      The average effect size of the selected singles.
+    """
+    all_singles = sampling.get_all_single_mutants(landscape.wildtype_sequence, landscape.vocab_size)
+    df = experiment.get_fitness_df(all_singles, landscape.evaluate, landscape.wildtype_sequence)
+    if adaptive:
+        mean_single_effect = df[df.fitness >= threshold].fitness.mean()
+    else:
+        mean_single_effect = df[df.fitness < threshold].fitness.mean()
+    return mean_single_effect
+
+
+def get_epistatic_horizon(landscape: potts_model.PottsModel) -> float:
+    r"""Returns the epistatic horizon for the given landscape.
+
+    The "epistatic horizon" is defined as the distance K from the wildtype at which, on average,
+    epistatic contributions outweigh linear contributions from adaptive singles. This is the average
+    distance we can expect a greedy algorithm to perform well on the given landscape.
+
+    Let $s_+$ be the average adaptive single mutant effect. let $e_{+,+}$ be the average
```

Comment:
Not sure if the latex dollar signs are useful here - usually docstrings don't get converted to markdown, right?

Reply:
They do not... but I'm not sure how else to indicate that this is an equation that uses subscripts, especially given the expressions below.